include/util.h

```diff
@@ -99,59 +99,131 @@ long CONST char_to_long(char c);
 long PURE str_to_long(const char *str);
 
 
-int __builtin_clzl(unsigned long x);
-int __builtin_ctzl(unsigned long x);
+#ifdef CONFIG_CLZ_NO_BUILTIN
+static unsigned clzl_impl(unsigned long x);
+static unsigned clzll_impl(unsigned long long x);
+#endif
 
-#ifdef CONFIG_ARCH_RISCV
-uint32_t __clzsi2(uint32_t x);
-uint32_t __ctzsi2(uint32_t x);
-uint32_t __clzdi2(uint64_t x);
-uint32_t __ctzdi2(uint64_t x);
+#if defined(CONFIG_CTZ_NO_BUILTIN) && defined(CONFIG_CLZ_NO_BUILTIN)
+static unsigned ctzl_impl(unsigned long x);
+static unsigned ctzll_impl(unsigned long long x);
 #endif
+
+// Used for compile-time constants, so should always use the builtin.
+#define CTZL(x) __builtin_ctzl(x)
+
+// Count leading zeros.
+#ifndef CONFIG_CLZ_NO_BUILTIN
+// If we use a compiler builtin, we cannot verify it, so we use the following
+// annotations to hide the function body from the proofs, and axiomatise its
+// behaviour.
+// On the other hand, if we use our own implementation instead of the builtin,
+// then we want to expose that implementation to the proofs, and therefore hide
+// these annotations.
 /** MODIFIES: */
 /** DONT_TRANSLATE */
 /** FNSPEC clzl_spec:
   "\<forall>s. \<Gamma> \<turnstile>
-    {\<sigma>. s = \<sigma> \<and> x_' s \<noteq> 0 }
+    {\<sigma>. s = \<sigma> \<and> x__unsigned_long_' s \<noteq> 0 }
       \<acute>ret__long :== PROC clzl(\<acute>x)
-    \<lbrace> \<acute>ret__long = of_nat (word_clz (x_' s)) \<rbrace>"
+    \<lbrace> \<acute>ret__long = of_nat (word_clz (x__unsigned_long_' s)) \<rbrace>"
 */
+#endif
 static inline long
 CONST clzl(unsigned long x)
 {
+#ifdef CONFIG_CLZ_NO_BUILTIN
+    return clzl_impl(x);
+#else
     return __builtin_clzl(x);
+#endif
 }
 
+#ifndef CONFIG_CLZ_NO_BUILTIN
+// See comments on clzl.
+/** MODIFIES: */
+/** DONT_TRANSLATE */
+/** FNSPEC clzll_spec:
+  "\<forall>s. \<Gamma> \<turnstile>
+    {\<sigma>. s = \<sigma> \<and> x__unsigned_longlong_' s \<noteq> 0 }
+      \<acute>ret__longlong :== PROC clzll(\<acute>x)
+    \<lbrace> \<acute>ret__longlong = of_nat (word_clz (x__unsigned_longlong_' s)) \<rbrace>"
+*/
+#endif
+static inline long long
+CONST clzll(unsigned long long x)
+{
+#ifdef CONFIG_CLZ_NO_BUILTIN
+    return clzll_impl(x);
+#else
+    return __builtin_clzll(x);
+#endif
+}
+
+// Count trailing zeros.
+#ifndef CONFIG_CTZ_NO_BUILTIN
+// See comments on clzl.
 /** MODIFIES: */
 /** DONT_TRANSLATE */
 /** FNSPEC ctzl_spec:
   "\<forall>s. \<Gamma> \<turnstile>
-    {\<sigma>. s = \<sigma> \<and> x_' s \<noteq> 0 }
+    {\<sigma>. s = \<sigma> \<and> x__unsigned_long_' s \<noteq> 0 }
       \<acute>ret__long :== PROC ctzl(\<acute>x)
-    \<lbrace> \<acute>ret__long = of_nat (word_ctz (x_' s)) \<rbrace>"
+    \<lbrace> \<acute>ret__long = of_nat (word_ctz (x__unsigned_long_' s)) \<rbrace>"
 */
+#endif
 static inline long
 CONST ctzl(unsigned long x)
 {
+#ifdef CONFIG_CTZ_NO_BUILTIN
+#ifdef CONFIG_CLZ_NO_BUILTIN
```

Comment:
could you explain the logic for the two #ifdefs here, better as comments in source code.

Reply:
See b5486e1.